File: flowman-tools/src/main/scala/com/dimajix/flowman/tools/exec/mapping/InspectCommand.scala
/*
 * Copyright 2021-2022 Kaya Kupferschmidt
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dimajix.flowman.tools.exec.mapping

import scala.util.control.NonFatal

import com.dimajix.flowman.execution.Context
import com.dimajix.flowman.execution.NoSuchMappingException
import com.dimajix.flowman.execution.Session
import com.dimajix.flowman.model.MappingIdentifier
import com.dimajix.flowman.model.Project
import com.dimajix.flowman.tools.exec.Command
import org.kohsuke.args4j.Argument
import org.slf4j.LoggerFactory

import com.dimajix.common.ExceptionUtils.reasons
import com.dimajix.flowman.execution.Status


class InspectCommand extends Command {
    private val logger = LoggerFactory.getLogger(classOf[InspectCommand])

    @Argument(required = true, usage = "specifies mapping to inspect", metaVar = "<mapping>")
    var mapping: String = ""

    override def execute(session: Session, project: Project, context: Context): Status = {
        try {
            val mapping = context.getMapping(MappingIdentifier(this.mapping))
            println("Mapping:")
            println(s"    name: ${mapping.name}")
            println(s"    kind: ${mapping.kind}")
            println(s"    inputs: ${mapping.inputs.map(_.toString).mkString(",")}")
            println(s"    outputs: ${mapping.outputs.mkString(",")}")
            println(s"    cache: ${mapping.cache.toString()}")
            println(s"    broadcast: ${mapping.broadcast}")
            println(s"    checkpoint: ${mapping.checkpoint}")
            println(s"  Requires:")
                mapping.requires
                    .map(_.text)
                    .toSeq.sorted
                    .foreach{ p => println(s"    $p") }
            Status.SUCCESS
        }
        catch {
            case ex:NoSuchMappingException =>
                logger.error(s"Cannot resolve mapping '${ex.mapping}'")
                Status.FAILED
            case NonFatal(e) =>
                logger.error(s"Error inspecting '$mapping':\n  ${reasons(e)}")
                Status.FAILED
        }
    }
}

File: flowman-kernel-tools/src/main/java/com/dimajix/flowman/tools/rexec/mapping/InspectCommand.java
/*
 * Copyright (C) 2023 The Flowman Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dimajix.flowman.tools.rexec.mapping;

import lombok.val;
import org.kohsuke.args4j.Argument;

import com.dimajix.flowman.kernel.model.MappingIdentifier;
import com.dimajix.flowman.kernel.model.MappingOutputIdentifier;
import com.dimajix.flowman.kernel.model.ResourceIdentifier;
import com.dimajix.flowman.kernel.model.Status;
import com.dimajix.flowman.tools.ExecutionContext;
import com.dimajix.flowman.tools.rexec.Command;


public class InspectCommand extends Command {
    @Argument(required = true, usage = "specifies mapping to inspect", metaVar = "<mapping>")
    String mapping = "";

    @Override
    public Status execute(ExecutionContext context) {
        val session = context.getSession();
        val mapping = session.getMapping(MappingIdentifier.ofString(this.mapping));
        System.out.println("Mapping:");
        System.out.println("    name: " + mapping.getName());
        System.out.println("    kind: " + mapping.getKind());
        System.out.println("    inputs: " + mapping.getInputs().stream().map(MappingOutputIdentifier::toString).reduce((l,r) -> l + "," + r));
        System.out.println("    outputs: " + mapping.getOutputs().stream().reduce((l,r) -> l + "," + r));
        System.out.println("    cache: " + mapping.getCache());
        System.out.println("    broadcast: " + mapping.isBroadcast());
        System.out.println("    checkpoint: " + mapping.isCheckpoint());
        System.out.println("  Requires:");
            mapping.getRequired()
                .stream()
                .map(ResourceIdentifier::toString)
                .sorted()
                .forEach(p -> System.out.println("    " + p));
        return Status.SUCCESS;
    }
}
